server/service/src/rnr_form/generate_rnr_form_lines.rs

```diff
@@ -1,95 +1,339 @@
+
+    let adjusted_quantity_consumed = match days_in_stock {
+        0 => 0.0,
+        days_in_stock => consumed * period_length_in_days as f64 / days_in_stock as f64,
+    };
+
+    adjusted_quantity_consumed
+}
+
+#[derive(Debug, PartialEq, Default, Copy, Clone)]
+pub struct UsageStats {
+    pub consumed: f64,
+    pub replenished: f64,
+    pub adjusted: f64,
+}
+
+pub fn get_usage_map(
+    connection: &StorageConnection,
+    store_id: &str,
+    item_id_filter: Option<EqualFilter<String>>,
+    period_length_in_days: i64,
+    end_date: &NaiveDate,
+) -> Result<HashMap<String, UsageStats>, RepositoryError> {
+    let lookback_days = period_length_in_days - 1; // period length is inclusive
+
+    let start_date = date_with_offset(end_date, Duration::days(lookback_days).neg());
+    let store_id_filter = Some(EqualFilter::equal_to(store_id));
+    let date_filter = Some(DateFilter::date_range(&start_date, end_date));
+
+    let consumption_rows =
+        ConsumptionRepository::new(connection).query(Some(ConsumptionFilter {
+            item_id: item_id_filter.clone(),
+            store_id: store_id_filter.clone(),
+            date: date_filter.clone(),
+        }))?;
+    let replenishment_rows =
+        ReplenishmentRepository::new(connection).query(Some(ReplenishmentFilter {
+            item_id: item_id_filter.clone(),
+            store_id: store_id_filter.clone(),
+            date: date_filter.clone(),
+        }))?;
+    let adjustment_rows = AdjustmentRepository::new(connection).query(Some(AdjustmentFilter {
+        item_id: item_id_filter,
+        store_id: store_id_filter,
+        date: date_filter,
+    }))?;
+
+    let mut usage_map = HashMap::new();
+
+    for consumption_row in consumption_rows.into_iter() {
+        let item = usage_map
+            .entry(consumption_row.item_id.clone())
+            .or_insert(UsageStats::default());
+        item.consumed += consumption_row.quantity;
+    }
+    for replenishment_row in replenishment_rows.into_iter() {
+        let item = usage_map
+            .entry(replenishment_row.item_id.clone())
+            .or_insert(UsageStats::default());
+        item.replenished += replenishment_row.quantity;
+    }
+    for adjustment_row in adjustment_rows.into_iter() {
+        let item = usage_map
+            .entry(adjustment_row.item_id.clone())
+            .or_insert(UsageStats::default());
+        item.adjusted += adjustment_row.quantity;
+    }
+
+    Ok(usage_map)
+}
+
+pub fn get_earliest_expiry(
+    connection: &StorageConnection,
+    store_id: &str,
+    item_id: &str,
+) -> Result<Option<NaiveDate>, RepositoryError> {
+    let filter = StockLineFilter::new()
+        .store_id(EqualFilter::equal_to(store_id))
+        .item_id(EqualFilter::equal_to(item_id))
+        // TODO: this is available stock _now_, but not what would have been available at the closing time of the period!
+        // Also, should it be available or in store?
+        .is_available(true);
```

Comment:
Is this ok? I guess would be a little strange to show currently available earliest expiring stock if you're filling forms in retroactively, but is that fairly unlikely?